tokio-fs/src/file/clone.rs

```diff
@@ -0,0 +1,37 @@
+use super::File;
+
+use futures::{Future, Poll};
+
+use std::io;
+
+const REPOLL_ERR: &str = "Cannot poll `FileCloneFuture` after it resolves";
+
+/// Future returned by `File::try_clone`.
+/// Clones a file handle into two file handles.
+/// Will panic if polled after returning an item or error.
+#[derive(Debug)]
+pub struct FileCloneFuture {
+    file: Option<File>,
+}
+
+impl FileCloneFuture {
+    pub(crate) fn new(file: File) -> Self {
+        Self { file: Some(file) }
+    }
+}
+
+impl Future for FileCloneFuture {
+    type Item = (File, File);
+    type Error = (File, io::Error);
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        self.file
+            .as_mut()
+            .unwrap()
+            .poll_try_clone()
+            .map(|async_| {
```

Comment:
nit: `poll_try_clone().map(|inner| inner.map(|cloned| (self.file.take().unwrap(), cloned_file)))` I would move the unwrap down here instead of above since that one will be hit first.